File: Algorithm/c_or_cpp/charpter8_tree_hot_problems/level1/is_symmetric.cpp
#include <vector>
#include <stack>
#include <iostream>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

   bool check(TreeNode *p, TreeNode *q) {
        if (!p && !q) return true;
        if (!p || !q) return false;
        if(p->val != q->val) return false; 
        return  check(p->left, q->right) && check(p->right, q->left);
    }

    bool isSymmetric(TreeNode* root) {
        return check(root, root);
    }

int main() {
    // 创建测试用例的二叉树
    TreeNode* root1 = new TreeNode(1);
    root1->left = new TreeNode(2);
    root1->right = new TreeNode(2);
    root1->left->left = new TreeNode(3);
    root1->left->right = new TreeNode(4);
    root1->right->left = new TreeNode(4);
    root1->right->right = new TreeNode(3);
  
    
    cout << isSymmetric(root1) << endl; // 输出1，表示对称
    return 0;
}

File: Algorithm/Leetcode/Code_Materials/python_source/charpter8_tree_hot_problems/level2/topic_双指针/is_symmetric.py
# coding=utf-8
from python_source.tool.Tree import Tree

#  判断树是否一样
class IsSymmetric:
    def isSymmetric(self, root):
        if not root:
            return True
        def dfs(left, right):
            # 递归的终止条件是两个节点都为空
            # 或者两个节点中有一个为空
            # 或者两个节点的值不相等
            if not (left or right):
                return True
            if not (left and right):
                return False
            if left.val != right.val:
                return False
            return dfs(left.left, right.right) and dfs(left.right, right.left)

        # 用递归函数，比较左节点，右节点
        return dfs(root.left, root.right)


if __name__ == "__main__":
    tree = Tree()

    root = tree.init_tree_for_symmetric()
    isSymmetric = IsSymmetric()
    res = isSymmetric.isSymmetric(root)
    print res
